device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,456 @@
+     * when a particular device's connection status changes, you will need to register a connection status change callback
+     * on that device client instance using {@link DeviceClient#registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback, Object)}.
+     *
+     * <p>Note that the thread used to deliver this callback should not be used to call open()/closeNow() on the client
+     * that this callback belongs to. All open()/closeNow() operations should be done on a separate thread</p>
+     *
+     * @param callback The callback to be fired when the connection status of the multiplexed connection changes.
+     *                 Can be null to unset this listener as long as the provided callbackContext is also null.
+     * @param callbackContext a context to be passed to the callback. Can be {@code null}.
+     */
+    public void registerConnectionStatusChangeCallback(IotHubConnectionStatusChangeCallback callback, Object callbackContext)
+    {
+        this.deviceIO.registerMultiplexingConnectionStateCallback(callback, callbackContext);
+    }
+
+    /**
+     * Returns if a device client for the provided device Id is already registered to this multiplexing client.
+     * @param deviceId The Id of the device client to look for.
+     * @return True if a device client is already registered with this Id. False otherwise.
+     */
+    public boolean isDeviceRegistered(String deviceId)
+    {
+        synchronized (this.operationLock)
+        {
+            for (DeviceClient client : this.deviceClientList)
+            {
+                if (client.getConfig().getDeviceId().equalsIgnoreCase(deviceId))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    /**
+     * Get the number of currently registered devices on this multiplexing client.
+     * @return The number of currently registered devices on this multiplexing client.
+     */
+    public int getRegisteredDeviceCount()
+    {
+        synchronized (this.operationLock)
+        {
+            // O(1) operation since ArrayList saves this value as an integer rather than iterating over each element.
+            // So there is no need to be more clever about this.
+            return this.deviceClientList.size();
+        }
+    }
+
+    /**
+     * Sets the given retry policy for the multiplexing client level connection management.
+     * <a href="https://github.com/Azure/azure-iot-sdk-java/blob/master/device/iot-device-client/devdoc/requirement_docs/com/microsoft/azure/iothub/retryPolicy.md">
+     *     See more details about the default retry policy and about using custom retry policies here</a>
+     * @param retryPolicy The policy that the multiplexing client will use when reconnecting.
+     */
+    public void setRetryPolicy(RetryPolicy retryPolicy)
+    {
+        this.deviceIO.setMultiplexingRetryPolicy(retryPolicy);
```

Comment:
I remember reading above that each device client can have its own retry policy. Would the device client level retry policy (if provided) override the multiplexing client level retry policy?

Reply:
No. Each device client's retry policy is only consulted when a message send on that particular device fails, or if that particular device has its session lost. For connection level events like TCP connection drop, only the multiplexing level retry policy is consulted

Reply:
We call it out as such in our javadoc comments, right?